### ISSUE TITLE
Incorrect sorting with sort criteria partially covered by index

### ISSUE THREAD

**Issue by mpoeter** [arangodb/arangodb#10440]
## My Environment

* __ArangoDB Version__:        3.4.8 & self-compiled devel branch
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Operating System__:        Ubuntu 18.04
* __Total RAM in your machine__:        16GB

## Component, Query & Data

__Affected feature__:
AQL query using web interface

__AQL query (if applicable)__:
```
FOR a in activities
    FILTER a.user_id == "31132147"
    SORT a.datetime DESC, a.order DESC, a._key DESC
    LIMIT 10
    RETURN KEEP(a, "_key", "name", "order", "datetime")
```

__AQL explain (if applicable)__:
```
Query String (178 chars, cacheable: true):
 FOR a in activities
     FILTER a.user_id == "31132147"
     SORT a.datetime DESC, a.order DESC, a._key DESC
     LIMIT 10
     RETURN KEEP(a, "_key", "name", "order", "datetime")

Execution plan:
 Id   NodeType          Est.   Comment
  1   SingletonNode        1   * ROOT
 12   IndexNode          589     - FOR a IN activities   /* reverse skiplist index scan */
  6   CalculationNode    589       - LET #5 = a.`order`   /* attribute expression */   /* collections used: a : activities */
  7   CalculationNode    589       - LET #7 = a.`_key`   /* attribute expression */   /* collections used: a : activities */
  8   SortNode           589       - SORT #5 DESC, #7 DESC
  9   LimitNode           10       - LIMIT 0, 10
 10   CalculationNode     10       - LET #9 = KEEP(a, "_key", "name", "order", "datetime")   /* simple expression */   /* collections used: a : activities */
 11   ReturnNode          10       - RETURN #9

Indexes used:
 By   Type       Collection   Unique   Sparse   Selectivity   Fields                                 Ranges
 12   skiplist   activities   false    false        37.29 %   [ `user_id`, `datetime`, `_key` ]   (a.`user_id` == "31132147")

Functions used:
 Name   Deterministic   Cacheable   Uses V8
 KEEP   true            true        false  

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   move-calculations-up-2
  4   move-filters-up-2
  5   use-indexes
  6   remove-filter-covered-by-index
  7   use-index-for-sort
  8   remove-unnecessary-calculations-2
```

__Problem__:
The query optimizer recognizes correctly that the filter and the first sorting criterion are covered by the index on `[user_id, datetime, _key]`, so it removes them. However, this still leaves the `SORT` node with the two remaining criteria and the query therefore produces a result that is not sorted correctly.

__Expected result__:
Result is sorted correctly

__Workaround__:
Create an index that covers all fields used for sorting.

## Steps to reproduce

1. Create a collection `activities`
2. Insert these sample documents [sample.json.gz](https://github.com/arangodb/arangodb/files/3847072/sample.json.gz)
3. Create an index for `[user_id, datetime, _key]`
4. Run the query

**Comment by jsteemann** [comment]
I think the dataset does not match the query somehow.
The dataset does not include any `user_id` attributes for any of the documents.

However, I will make up some data on my own and try to reproduce.

**Comment by jsteemann** [comment]
Reproduced. Looking into it...

**Comment by mpoeter** [comment]
Right, sorry. I simply took the result of the query and forgot that it doesn't include a user_id.

**Comment by jsteemann** [comment]
The above PRs address this issue.
They are expected to land in
* 3.4.9: https://github.com/arangodb/arangodb/pull/10444
* 3.5.4: https://github.com/arangodb/arangodb/pull/10443
* devel: https://github.com/arangodb/arangodb/pull/10442

**Comment by maxkernbach** [comment]
Hi @mpoeter,

Version 3.4.9, which includes a fix for the sorting, has just been released.
From the [Changelog](https://github.com/arangodb/arangodb/blob/9a2aa342442dbe0d1d9f3918642e30ba9317ad02/CHANGELOG#L58):
> * Fixed issue #10440: Incorrect sorting with sort criteria partially covered
  by index.

In case this issue still occurs after upgrading, please comment and we will reopen the ticket.

Please find the latest release in the [download section](https://www.arangodb.com/download/).
